Update persisted queries docs (#3160)
apollographql/apollo-ios: docs/source/fetching/persisted-queries.mdx

- using the [`generate-operation-manifest`](./../code-generation/codegen-cli#generate-operation-manifest) command; this generates the operation manifest without running full code generation

The resulting operation manifest for `persistedQueries` looks something like this: 

```json title="operationIdentifiers.json"
{
  "format": "apollo-persisted-query-manifest",
  "version": 1,
  "operations": [
    {
      "id": "e0321f6b438bb42c022f633d38c19549dea9a2d55c908f64c5c6cb8403442fef",
      "body": "query GetItem { thing { __typename } }",
      "name": "GetItem",
      "type": "query"
    }
  ]
}
```

The resulting operation manifest JSON for `legacyAPQ` looks something like this: 

```json title="operationIdentifiers.json"
{
  "525cd90917f8945d8d6a68fde991aaf766ff7d7b227ab2cd8dc97b7959c8d0d6" : {
    "name": "GetItem",
    "source": "query GetItem { thing { __typename ...Name } } fragment Name on Item { __typename name }"
  }
}
```

To automatically update the manifest for each new app release, you can include the [`generate`](./../code-generation/codegen-cli#generate) command in your CI/CD pipeline.

> #### Publish manifests to the PQL

Reviewer: I don't think this section should be part of this doc since it doesn't really pertain to the iOS portion. At most I think if we want to mention/reference the PQL we should link out to a separate docs page for that.

Reviewer 2: Hmm, I think it's preferable to have one place where the publish instructions are referenced for ease of maintenance but it's a pretty small section so I'm not completely opposed to having it here.

Author: I wrestled with this too, since this isn't iOS specific. At the same time, if I was a client maintainer, I'd prefer to have everything in one place, which is why I decided to put it in. If others feel strongly about removing it, we can do that.

Author: > Is this being duplicated into the Kotlin docs too?
Yes! I'd like to have parity between all client docs on this.

BTW, it'll also be in the ["main PQ" docs](https://github.com/apollographql/docs/pull/496) regardless, just want to make it easier for a client dev by having everything here. Am considering making reusable stubs for any sections that will be duplicated.

Reviewer 3: I'd like to keep this section as well. It would be nice if we could share this block across the different platforms documentation though.

I know we have a folder of `shared` `.md` files that can be placed in-line inside other documentation pages in the iOS repo. But could we share this across the different docs platforms? That way we guarantee we're all using the same verbiage and it can be kept up to date in a single place. @trevorblades Is that possible?

Author: Good question, @AnthonyMDev ! I didn't think about/realize that these shared `.md` were on per-repo basis.
Maybe easiest just to make a new React component with just text in the main docs repo? Let me know what you think @trevorblades

Reviewer 4: https://github.com/apollographql/docs/pull/515 exposes all of these shared bits of MDX as components that you can use in your docs' content:

```mdx
- some
- markdown
- content

<ObtainApiKey />
```

If you want to see the list of components that can be used, you can check out the export names in the `src/content/shared/index.js` file in the docs repo.

Author: Yay, ty!!!! Will replace text sections in client docs with these tomorrow.

---
title: Persisted Queries
---

Apollo supports two separate but related features called **persisted queries** and **automated persisted queries** (APQs).
With both features, clients can execute a GraphQL operation by sending an operation's ID instead of the entire string. An operation's ID is the SHA256 hash of the full operation string. Querying by ID can significantly reduce latency and bandwidth usage for very large operation strings.

## Differences between persisted queries and APQ

The persisted queries feature requires operations to be preregistered in a **persisted query list** (**PQL**). 
This allows the PQL to act as a safelist of trusted operations made by your first-party apps. As such, persisted queries is a security feature as much as a performance one.

> With APQs, if the server can't find the operation ID the client provides, the server returns an error indicating that it needs the full operation string. If the Apollo iOS client receives this error, it automatically retries the operation with the full operation string.

Reviewer 2: > If the Apollo iOS client receives this error, it automatically retries the operation with the full operation string.

I don't remember what the default for APQ behaviour is in Apollo iOS, but this may not be true if it's disabled by default.

Author: Ah, good to know. @BobaFetters can you confirm the default behavior?

Will reword if the retries are disabled by default.

Reviewer: From what I see in the code it looks like if you have `autoPersistQueries` turned on in the networking then it will automatically retry the request if it fails if the error message contains `PersistedQueryNotFound`, and some other checks like to make sure it hasn't retried already etc

Author: Thanks for digging! Is this configurable at all?

Reviewer: Doesn't appear to be, looks like if you turn on auto persisted queries the retry based on the `PersistedQueryNotFound` error will always be how this is handled.

With APQs, if the server can't find the operation ID the client provides, the server returns an error indicating that it needs the full operation string. If the Apollo iOS client receives this error, it automatically retries the operation with the full operation string.

If you _only_ want to improve request latency and bandwidth usage, APQ addresses your use case. If you _also_ want to secure your supergraph with operation safelisting, you should preregister trusted operations in a PQL.

For more details on differences between persisted queries and APQ, see the [GraphOS persisted queries documentation](/graphos/operations/persisted-queries#differences-from-automatic-persisted).

### Requirements

Persisted queries is currently in [preview](/resources/product-launch-stages#preview) and has the following requirements:

> - Apollo iOS (v1.3.1+)

Reviewer: ```suggestion
- Apollo iOS (v1.4.0+)
```

Author: Ah, thanks for calling that out!

- [GraphOS Enterprise plan](/graphos/enterprise/)

You can use APQ with the following versions of Apollo iOS, Apollo Router, and Apollo Server:
- Apollo iOS (v1.0.0+)

> - [Apollo Router](/router) (v0.1.0+)

Author: Great feedback! Looks like Router supports APQ, based on this [Slack thread](https://apollograph.slack.com/archives/C04LHRSTLDD/p1690816533569319).

It makes sense you could use APQ on Server without the router, so will clarify that.

- [Apollo Server](/apollo-server/) (v1.0.0+)

## Implementation steps

> Both persisted queries and APQ require you to configure code generation and how your client makes requests. The rest of this article details these configurations.

Reviewer 5: Nit: only only PQs require to generate a manifest. APQs work out of the box, at least on Android, there's no codegen configuration to do at all

## Implementation steps

Both persisted queries and APQs require you to configure code generation and how your client makes requests. If you intend to use persisted queries for safelisting, you also need to generate an operations manifest.

We recommend you follow this order while implementing: 

> 1. **For persisted queries only**: Generate and publish operation manifests

Reviewer 3: ```suggestion
1. **For persisted queries only**: Generate and publish an operation manifest
```

> We recommend you follow this order while implementing: 
> 
> 1. **For persisted queries only**: Generate and publish operation manifests
> 2. **For both persisted queries and APQs**: Configure generated operation models
> 3. **For both persisted queries and APQs**: Enable persisted queries on `ApolloClient`

Reviewer 3: There is no reason generating and publishing the manifest needs to happen first. And I think it would be clearer that it's only necessary for persisted queries if we put it last like this:

```suggestion
To implement these features, follow these steps.

1. Configure generated operation models
2. Enable persisted queries on `ApolloClient`
3. **For persisted queries only**: Generate and publish operation manifests
```

And then move the "Generate Operation Manifests" section to the bottom.

Reviewer 3: I'm recognizing that this has been added as one of the pieces that we are going to replace with shared components . So if this suggested change is preferred by others, we should make the change there and then use that component.

Author: I think this order makes the most sense,  even if it sticks the PQ-only step in the middle.
1 - Code generation
2 - Publish operation manifest
3 - Client updates 

(It also matches the implementation narrative in the "[main](https://github.com/apollographql/docs/pull/496/files)" PQ docs, though I'm open to changing it if it's not the most user-intuitive.)

If the sole purpose of putting manifest publication last is to make clearer which step is PQ only, I'd say it'd be better to find another way to make this clearer.

> 1. In the `output` section of your configuration, set the [`operationManifest`](./../code-generation/codegen-configuration#operation-manifest) with:

Reviewer 3: If you want to generate operation manifest every time you do codegen, this is how you do that, but actually, we recommend that you only generated the manifest as needed using the CLI command as noted below on line 67 of this document.

It looks like we actually need to make an addition to the `generate-operation-manifest` command to make this possible though. So this documentation is accurate to the current state of things. Speaking with @BobaFetters about making the addition so we can update these docs to reflect what we want users to be able to do.

---

You really only need the `operationManifest` for `APQ` if you are also using the safe-listing capabilities or choosing to pre-register your operations with APQs. 

The messaging up above explaining the difference between `APQ` and `persistedQueries` doesn't mention anything about pre-registering documents when using `APQs. It seems to indicate that pre-registering queries is really only a feature of `persistedQueries`.

@glasser Do we still want to even be mentioning the ability to pre-register queries in the APQ world? Or are we trying to remove references to that feature and discourage its use since it was buggy and `persistedQueries` is replacing it?

---

Once we've made the change, I'm wondering if we could make this more clear by organizing this as follows:

It could be 

```markdown
### 1. Generate an operation manifest
{Explain what the operation manifest is for.}
{Explain that the manifest is needed for persisted queries and optional for APQs if you want to pre-register them?}

{Explain the primary way of generating operation manifest, via the CLI command with the `--path` and `--version` arguments}

{Explain the alternative way of adding the `output.operationManifest` to codegen config. Which will re-generate the manifest everytime you run the `generate` command.}

### 2. Configure generated operation models
{Insert the section on `options.operationDocumentFormat` from lines 54-63.}

### 3. Enable persisted queries on `ApolloClient`
{Insert Existing Section}
```

Reviewer 3: PR #3172 is up which adds the desired functionality. We should now change these docs as indicated above.

Author: > Or are we trying to remove references to that feature and discourage its use since it was buggy and persistedQueries is replacing it?

I'd imagine we'd prefer folks to use `persistedQueries` over APQs, so will keep the intro as it is. (I.e. not mention you _can_ preregister queries with APQ.)

Re: removing all mentions of `apq`... not sure. Will bring up with product / PMM.

Reviewer 5: > Do we still want to even be mentioning the ability to pre-register queries in the APQ world?

I would vote against mentioning that. I actually don't think you need to pre-register anything with APQ. APQ are completely standalone and don't need pre-registering. This is the beauty of it. On Kotlin, we have an option to generate a manifest in the "old apollo-cli format" (we call it operationOutput) but to my knowledge this isn't used at all by APQ. This was for users who implemented their own persisted queries manually.

The rest of this article details these steps.

> Persisted queries also requires you to create and link a PQL, and to configure your router. For more information on the other configuration aspects of persisted queries, see the [GraphOS persisted queries documentation](/graphos/operations/persisted-queries).

Reviewer 3: ```suggestion
This document only describes the steps that need to be taken by the client to send persisted query requests and create a manifest of the client's operations. Persisted queries also require you to create and link a PQL, and to configure your router to receive persisted query requests. For more information on the other configuration aspects of persisted queries, see the [GraphOS persisted queries documentation](/graphos/operations/persisted-queries).
```

> - the `version`: either `persistedQueries` for persisted queries or `legacyAPQs` for APQs

Reviewer 3: Before making this feature stable, should we be naming this `legacyAPQs` or just `APQs`? I was under the assumption that once `persistedQueries` was public availability, that the `APQ` feature would be deprecated across the entire platform. But is that assumption correct? @glasser

Reviewer 5: I like `APQ`! They are very easy to setup, we use them in [Confetti](https://github.com/joreilly/Confetti). I would be sad to see them go but I'll follow what's decided here.

Reviewer 6: @glasser what's your read on this?

Reviewer 5: My read is that APQs are still a thing and not going anywhere. The "old apollo cli" manifest format on the other hand can be deprecated. I would link the "old apollo cli" manifest format to APQ. 

In Apollo Kotlin we called that format `operationOutput` (I think I can remember the apollo cli using something like that but I wouldn't bet on it). But any format like `legacyManifest` would do.

Reviewer 3: So does that mean that the new manifest format can still be used to persist queries that work with APQs? Has apollo server been updated to support this?

Reviewer 3: To be clear, I understand that APQs _generally_ don't need the manifest, since they are automatically-persisted. I'm talking specifically about users that have been using the existing APQ functionality to work as a safelisting mechanism by pre-registering the operations with a manifest.

Reviewer 7: I don't know much about the uses of the older manifest format here — was it for the legacy operation registry/safelisting feature (which was not connected in any way to APQs)? We will certainly be eventually shutting that feature off (it has been undocumented and barred to new users for a while now).

We're not intending to deprecate APQs, which are an easy to set up performance feature (but don't provide any security benefits).

Reviewer 3: Okay, so I was under the impression that the legacy safelisting feature was connected to APQs. If it will be eventually shut off, then we can rename this to just `legacy` and move forward. Thanks @glasser

    "operationManifest" : {
        "path" : "./generated/operationIdentifiers.json",
        "version" : "persistedQueries"
    }
}
```

> 

Reviewer 3: I'd like a little better explanation of what these options do. Could we add here:

> Setting the `output.operationManifest` option will cause the operation manifest to be generated every time you run code generation.

> 2. In the `options` section of your configuration, set the `operationDocumentFormat` array to `definition`, `operationId`, or both `definition` and `operationId`. This determines how to generate the operation documents for your generated operations.

Reviewer 3: It needs to be stated here that for `APQs` to work, you **MUST** include both the `definition` and `operationId`. For `persistedQueries`, you only need the `operationId`, the `definition` is superflous, since the definition is already persisted.

  "version": 1,
  "operations": [
    {
      "id": "e0321f6b438bb42c022f633d38c19549dea9a2d55c908f64c5c6cb8403442fef",
      "body": "query GetItem { thing { __typename } }",
      "name": "GetItem",
      "type": "query"
    }
  ]
}
```

The resulting operation manifest JSON for `legacyAPQ` looks something like this: 

```json title="operationIdentifiers.json"
{
  "525cd90917f8945d8d6a68fde991aaf766ff7d7b227ab2cd8dc97b7959c8d0d6" : {
    "name": "GetItem",
    "source": "query GetItem { thing { __typename ...Name } } fragment Name on Item { __typename name }"
  }
}
```

To automatically update the manifest for each new app release, you can include the [`generate`](./../code-generation/codegen-cli#generate) or [`generate-operation-manifest`](./../code-generation/codegen-cli#generate-operation-manifest) command in your CI/CD pipeline.

#### Publish manifests to the PQL

<blockquote>

**This step is only required for persisted queries, not APQ.**

Ensure your Rover CLI version is `0.17.2` or later. Previous versions of Rover don't support publishing operations to a PQL.
[Download the latest version.](/rover/getting-started/)

</blockquote>

> If you want to preregister your manifest as a safelist of trusted operations, you need to publish it to a PQL using the [Rover CLI](/rover/):

Reviewer 5: Any interest in doing that with a Gradle task to avoid downloading an extra tool? Should be relatively easy to do if we want to.

Author: @AnthonyMDev @BobaFetters , thoughts on this?

Reviewer 3: Kotlin can do this with a Gradle task, but the iOS build system does not have Gradle or an equivalent tool that is easily available. Using the Rover CLI is the best approach for now on iOS.

> If you want to preregister your manifest as a safelist of trusted operations , you need to publish it to a PQL using the [Rover CLI](/rover/):

Reviewer 3: ```suggestion
If you want to preregister your manifest as a safelist of trusted operations, you need to publish it to a PQL using the [Rover CLI](/rover/):
```

```bash title="Example command"
rover persisted-queries publish my-graph@my-variant \
  --manifest ./persisted-query-manifest.json
```

> - Replace `my-graph@my-variant` with the **graph ref** of _any_ variant your PQL is [linked to](#12-link-the-pql-to-variants). Your platform team can provide this for you.

Reviewer 3: `[linked to](#12-link-the-pql-to-variants)`

Looks like this is supposed to link to an anchor in another page that this line was copied from. But this isn't going to be a valid link.

I think this stuff is all the more reason to put this section in a shared block referenced by all the platforms if possible.

Author: Very true, thanks for catching.

- Apollo iOS (v1.0.0+)
- [Apollo Server](/apollo-server/) (v1.0.0+)
- [Apollo Router](/router) (v0.1.0+)

> **Note:** You can use _either_ Apollo Server _or_ Apollo Router for APQs. They don't need to be used together.

## Implementation steps

Both persisted queries and APQs require you to configure code generation and how your client makes requests. If you intend to use persisted queries for safelisting, you also need to generate an operations manifest.

We recommend you follow this order while implementing: 

1. **For persisted queries only**: Generate and publish operation manifests
2. **For both persisted queries and APQs**: Configure generated operation models
3. **For both persisted queries and APQs**: Enable persisted queries on `ApolloClient`

The rest of this article details these steps.

Persisted queries also requires you to create and link a PQL, and to configure your router. For more information on the other configuration aspects of persisted queries, see the [GraphOS persisted queries documentation](/graphos/operations/persisted-queries).

### 1. Generate operation manifests

> > This step is only required for implementing safelisting with persisted queries, it is _not_ for required APQs.

Reviewer 7: comma splice, and I think "required" is in the wrong place?

Author: https://github.com/apollographql/apollo-ios/pull/3160/commits/99b7a92a8400e606052f4d1e9b225940660e3f27

An operation manifest acts as a safelist of trusted operations the [Apollo Router](/router/) can check incoming requests against. You can generate operation manifests in two ways:

> - by running the [`generate-operation-manifest`](./../code-generation/codegen-cli#generate-operation-manifest) CLI command with `--path` and `--version` arguments

Reviewer 7: Should this say subcommand?

- Initialize a custom `NetworkTransport` using `RequestChainNetworkTransport` with `autoPersistQueries` parameter set to `true` and an `interceptorProvider` that includes the `AutomaticPersistedQueryInterceptor` (such as `DefaultInterceptorProvider`).
- Initialize your `ApolloClient` with the custom `NetworkTransport` that supports persisted queries.

```swift
let store = ApolloStore(cache: InMemoryNormalizedCache())
let interceptorProvider = DefaultInterceptorProvider(store: store)
let networkTransport = RequestChainNetworkTransport(
    interceptorProvider: interceptorProvider,
    endpointURL: URL(string: "http://localhost:4000/graphql")!,
    autoPersistQueries: true
)

let client = ApolloClient(networkTransport: networkTransport, store: store)
```

For more information on configuring your `ApolloClient`, `NetworkTransport`, `InterceptorProvider`, and the request chain, see the [request pipeline](./../networking/request-pipeline) documentation.

#### Sending APQ retries as `GET` requests

> **Note:** Apollo iOS only retries failed ID-based operations for APQs, not persisted queries.

> By default, the Apollo clients sends operation retries as `POST` requests.  In some cases, you may prefer or need to retry an operation using a `GET` request: for example, you may make requests to a CDN that has better performance with `GET`s.

Reviewer 7: This isn't about the docs, just the feature — I'm a bit confused as to why iOS can be configured to use GETs for *retries* specifically, vs for all concise requests?

Reviewer 5: I think concise requests are always `GET` by default, full requests are always `POST` by default.

So if you want to change your full requests to `GET` you can. Not sure who does this but we have the same configuration option on Apollo Kotlin. I'd be OK leaving it out of the documentation if we think it's not that useful.

---
title: Persisted Queries

> description: Reject unrecognized operations while minimizing request latency

Reviewer 5: (absolutely optional) suggestion. Mention the "what" instead of the "how" in the subtitle? 
IMO the subtitle should be about the "what do I want to achieve" to invite the reader to read the "how do I achieve it". 

```suggestion
description: Secure your graph while minimizing request latency
```

Author: This version is better, thanks @martinbonnin !

For more details on differences between persisted queries and APQ, see the [GraphOS persisted queries documentation](/graphos/operations/persisted-queries#differences-from-automatic-persisted).

### Requirements

Persisted queries is currently in [preview](/resources/product-launch-stages#preview) and has the following requirements:
- Apollo iOS (v1.3.1+)
- [Apollo Router](/router) (v1.25.0+)
- [GraphOS Enterprise plan](/graphos/enterprise/)

You can use APQ with the following versions of Apollo iOS, Apollo Server, and Apollo Router:
- Apollo iOS (v1.0.0+)
- [Apollo Server](/apollo-server/) (v1.0.0+)
- [Apollo Router](/router) (v0.1.0+)

> **Note:** You can use _either_ Apollo Server _or_ Apollo Router for APQs. They don't need to be used together.

## Implementation steps

Both persisted queries and APQs require you to configure code generation and how your client makes requests. If you intend to use persisted queries for safelisting, you also need to generate an operations manifest.

We recommend you follow this order while implementing: 

> 1. **For persisted queries only**: Generate and publish operation manifests

Reviewer 5: Nitpick: from a client point of view, there is a single manifest so I would use singular form here. Of course users with multiple schemas would have multiple manifest but I think it's an edge case and we should optimize for the majority.

> 2. **For both persisted queries and APQs**: Configure generated operation models

Author: This is probably because I'm unfamiliar with code generation, but this step label/title confuses me. Is it enough to say "Configure code generation"?

Reviewer 3: I like "Configure generated operation models" more!

3. **For both persisted queries and APQs**: Enable persisted queries on `ApolloClient`

The rest of this article details these steps.

Persisted queries also requires you to create and link a PQL, and to configure your router. For more information on the other configuration aspects of persisted queries, see the [GraphOS persisted queries documentation](/graphos/operations/persisted-queries).

### 1. Generate operation manifests

> This step is only required for implementing safelisting with persisted queries, it is _not_ for required APQs.

An operation manifest acts as a safelist of trusted operations the [Apollo Router](/router/) can check incoming requests against. You can generate operation manifests in two ways:
- by running the [`generate`](./../code-generation/codegen-cli#generate) or [`generate-operation-manifest`](./../code-generation/codegen-cli#generate-operation-manifest) CLI command with `--path` and `--version` arguments

>     - `path` is where to write the manifest file

Author: I assumed that the `path` and `version` arguments would be used in the same was as they are in the config.json, but looking at the [`generate-operation-manifest`](https://www.apollographql.com/docs/ios/code-generation/codegen-cli#generate-operation-manifest) docs, it looks like `path` is where to read the config from (so you have to update the config anyway?) and the `version` option just spits out the CLI version?

Reviewer: We had some discussions around the CLI command and config json yesterday and there are going to be some changes to it that I am working on today so this documentation will be updated some based on that and hopefully everything will be more clear.

Author: Thank you!
